lib/service/service.go

```diff
@@ -4798,6 +4803,11 @@ func (process *TeleportProcess) StartShutdown(ctx context.Context) context.Conte
 	warnOnErr(process.closeImportedDescriptors(""), process.log)
 	warnOnErr(process.stopListeners(), process.log)
 
+	// populate context values
+	if len(process.getForkedPIDs()) > 0 {
```

Comment:
This only catches the across-process restart on SIGHUP, but doesn't catch the in-process restart on CA rotations; instead of checking this here, can we just inject the `ProcessForkedContext` (which should be `ProcessRestartingContext`, maybe) in the two places in which we call `StartShutdown` for restarts?

Reply:
> This only catches the across-process restart on SIGHUP

The [upgrade guide](https://goteleport.com/docs/management/admin/upgrading-the-teleport-binary/) instructs a `SIGUSR2`, then followed by a `SIGQUIT` when the child is up. So I was thinking the in-process restart can be added separately with a different context value.

Though now you mention it, it does feel easier to consolidate them into one. Is this the only place for in-process restart?
https://github.com/gravitational/teleport/blob/2fe049e6c3d0b1438eef6b0d6493fef3ee1f2bd3/lib/service/service.go#L735-L738

Reply:
You're actually right, we ought to check both possibilities, as the shutdown on SIGQUIT after a SIGUSR2 would just be a normal shutdown (not that SIGUSR2 can ever be used in a sensible way, mind you).

I believe that's the only place where we do in-process restarts. Keeping them separate would help because eventually we should get rid of in-process restarts altogether (host CA rotations/credentials reissuance is the only reason we still have that, but it should be possible to swap out credentials on the fly) and only SIGHUP/SIGUSR2 would remain.

Reply:
cool. then I will merge this PR as it is and I will do a separate PR for the in-process restart and test out the rotation scenario for #4422